packages/google_maps_flutter/lib/src/google_map.dart

```diff
@@ -135,36 +141,97 @@ class _GoogleMapState extends State<GoogleMap> {
   void initState() {
     super.initState();
     _googleMapOptions = _GoogleMapOptions.fromWidget(widget);
+    _markerUpdateHandler = _MarkerUpdateHandler.init(null);
   }
 
   @override
   void didUpdateWidget(GoogleMap oldWidget) {
     super.didUpdateWidget(oldWidget);
+    _updateOptions();
+    _updateMarkers();
+  }
+
+  void _updateOptions() async {
     final _GoogleMapOptions newOptions = _GoogleMapOptions.fromWidget(widget);
     final Map<String, dynamic> updates =
         _googleMapOptions.updatesMap(newOptions);
-    _updateOptions(updates);
-    _googleMapOptions = newOptions;
-  }
-
-  void _updateOptions(Map<String, dynamic> updates) async {
     if (updates.isEmpty) {
       return;
     }
     final GoogleMapController controller = await _controller.future;
     controller._updateMapOptions(updates);
+    _googleMapOptions = newOptions;
+  }
+
+  void _updateMarkers() async {
+    final GoogleMapController controller = await _controller.future;
+    controller._updateMarkers(_markerUpdateHandler.update(widget.markers));
   }
 
   Future<void> onPlatformViewCreated(int id) async {
-    final GoogleMapController controller =
-        await GoogleMapController.init(id, widget.initialCameraPosition);
+    final GoogleMapController controller = await GoogleMapController.init(
+      id,
+      widget.initialCameraPosition,
+      _markerUpdateHandler._onMarkerTap(),
+      _markerUpdateHandler._onInfoWindowTap(),
+    );
     _controller.complete(controller);
     if (widget.onMapCreated != null) {
       widget.onMapCreated(controller);
     }
   }
 }
 
+class _MarkerUpdateHandler {
```

Comment:
I find the role of this class a little hard to grasp, it does the following:
 * Maintains the set of current markers
 * Proxy calls to `MarkerUpdates` for computing and returning the diff for a new markers set.
 * Provides closures for invoking the marker events on a given marker.

I was also a little confused by the name - the initial mental model I had just from seeing the class name was (this is a marker update handler, it probably handles computing the updates and sending them over the platform channel).

The closure generation is particularly a trick I'd rather avoid if it's not necessary as it is harder to follow.

It might be easier to follow if we keep the markers map as a member of _GoogleMapState, then we could avoid the closures as we could have `_onMarkerTap(String markerId)` as a method of _GoogleMapState.
It looks like the update computation logic can be done in a pretty short method which we can fit in `_updateMarkers`.